packages/xstate-wallet/src/store/memory-store.ts

```diff
@@ -220,29 +234,30 @@ export class MemoryStore implements Store {
       _.pick(stateVars, 'outcome', 'turnNum', 'appData', 'isFinal')
     );
 
-    return Promise.resolve(entry);
+    return entry;
   }
-  private getNonce(addresses: string[]): BigNumber {
-    return this._nonces[this.nonceKeyFromAddresses(addresses)] || bigNumberify(-1);
+  private async getNonce(addresses: string[]): Promise<BigNumber> {
+    const nonce = await this.backend.getNonce(this.nonceKeyFromAddresses(addresses));
+    return nonce || bigNumberify(-1);
   }
 
-  private setNonce(addresses: string[], value: BigNumber) {
-    if (value.lte(this.getNonce(addresses))) throw 'Invalid nonce';
-
-    this._nonces[this.nonceKeyFromAddresses(addresses)] = value;
+  private async setNonce(addresses: string[], value: BigNumber) {
+    const nonce = await this.getNonce(addresses);
+    if (value.lte(nonce)) throw 'Invalid nonce';
+    await this.backend.setNonce(this.nonceKeyFromAddresses(addresses), value);
```

Comment:
This is unsafe async code: It could be that on L246, the nonce is valid. But then, on line 247, the nonce is no longer valid.

The backend needs to be responsible for ensuring nonce safety, and it would thus throw on line 247 in that case, making L246 redundant.

Reply:
Umm... okay.
The solution would be to set it like this way?
```suggestion
    if (value.lte(await this.getNonce(addresses))) throw 'Invalid nonce';
    await this.backend.setNonce(this.nonceKeyFromAddresses(addresses), value);
```

Or I could move this entire method to the backend, but even then, the lookup to the indexedDB would render that move useless. 
The thing is, at some point, we will always have some asynchronicity if we use an indexedDB storage to store nonces.

Reply:
Yes, the point is the lookup to the backend is useless: we should just ask the backend to set the nonce, and expect that it might throw an error:
```
try {
  await this.backend.setNonce(this.nonceKeyFromAddresses(addresses), value);
} catch (e) {
  // ...
}
```

Reply:
So maybe we should have a method `getNextNonce(key)` on the `DBBackend`, which is responsible for returning a new, unique nonce?

Depending on the transaction capabilities of IndexedDB (which are slightly unclear to me), we might be forced to have some unsafe async code. It would be good to isolate this as much as possible (i.e. into the `DBBackend`) though.

Reply:
The `setNonce` method would not be used by the creator of the channel; if you create a channel and ask me to join it, I have to `setNonce` with the nonce you've proposed.

Reply:
> 

Yes, the point is the lookup to the backend is useless: we should just ask the backend to set the nonce, and expect that it might throw an error:

try {
  await this.backend.setNonce(this.nonceKeyFromAddresses(addresses), value);
} catch (e) {
  // ...
}

I think that if the problem is the uniqueness of the nonce, we could make indexedDB ensure the uniqueness of the key. On indexedDB we can set a key to be used as the lookup key, and an optional rule, is to make that key unique.

Reply:
@andrewgordstewart @tomclose I've got a CRUD working now of IndexedDB, but I'm not clear on what should I do (if anything) about this.